### DISCUSSION TITLE
Refreshing selectors causes re-render before new value is computed

### DISCUSSION THREAD

**Discussion by sazzer** [facebookexperimental/Recoil#1948] [Q&A] [answered]
I've recently started using Recoil, and am impressed so far.

One issue that I've just hit though is the way rendering and refresh works together.

I've got a selector to store the current user details - loaded from the API. This is used in two different areas - to show the current users name in the top corner, and to display the user profile page. And when loading the user profile page it forces a refresh to ensure the latest version is showing. All good so far.

Except that when the profile page loads, the refresh causes the two different places to re-render *before* the new value is displayed. Which means they fall back to the `<Suspense />` closest to them. Which, in turn, means that the top corner name reverts to a Loading spinner until the API call returns.

This happens regardless of how I force the selector to refresh. In particular, I've tried:
* useRecoilRefresher_UNSTABLE
* Using an atom to store a Last Refreshed date, and updating that

I've also tried replacing the use of `useRecoilValue` with `useRecoilValue_TRANSITION_SUPPORT_UNSTABLE` and it's made no difference.

Is there any way to make selectors update, but only have components render the values when the new one is finally loaded instead of reverting back to the unloaded state?

Cheers

**Comment by drarmstr** [comment, marked as answer]
If the value is being updated to an async value then there is the possibility that React will render the pending state before the final state is rendered.  Options to deal with this include locating a `<Suspense>` boundary around the relevant section with an appropriate fallback, using a [React transition](https://recoiljs.org/docs/guides/transitions) in conjunction with `useRecoilValue_TRANSITION_SUPPORT_UNSTABLE()` to render the previous state while it is pending, or debouncing and only updating the state with the new computed state instead of refreshing or setting to an async Promise.

**Reply by sazzer** [reply]
So I get that when the selector is *first* loaded, React will need to render something before the value is computed. That's reasonable.

The bit I've been struggling with - and I'm *not* an expert React user! - is the same when the selector already has a value and it's being refreshed.

So, for example, the following flow:
1. Load page
2. Log in
    1. Selector is queried for the first time
    2. Selector makes API call to get current user details
3. Open User Profile screen
    1. A `useEffect` on the screen causes the selector to be refreshed
    2. Selector makes API call to get current user details again

It makes perfect sense during 2.2 that there's no value to display. It makes less sense during 3.2, because we already have a (potentially stale) value to display.

Interestingly, just wrapping the call to refresh within the `useEffect()` in a `startTransition()` *seems* to have done the job :)

**Comment by drarmstr** [comment]
If refreshing the selector causes it to query again and thus return an async Promise to a value then you are essentially first putting it in a pending state and then when that resolves it goes into a final state.  React may render the intermediate pending state.  Stale values are not rendered unless using something like [React transitions](https://recoiljs.org/docs/guides/transitions).  That's what wrapping in the `startTransition()` will help do.  So, glad they worked for you!

**Reply by salvoravida** [reply]
Hi all @drarmstr @sazzer, I used to solve this problem by saving in a ref the previous value with my easy wrapper
useRecoilQuery -> https://github.com/salvoravida/recoil-toolkit/blob/master/packages/recoil-toolkit/src/_core/hooks/useRecoilQuery.ts#L19

That said, I will suggest adding a new field to the Loadable, so when the selector is in the state "loading" a "previousContents" may contain the previous resolved value if any, instead of undefined.

I used to show "...loading...." on my upcoming DevTools! 
<img width="1188" alt="devtool-recoil-005" src="https://user-images.githubusercontent.com/20126259/187123997-3f6e3a68-0329-4b1a-b041-ce88def23efe.png">

**Reply by drarmstr** [reply]
@salvoravida Haha, yes, I've also used the "save in ref" hack!  However, I don't recommend it as a general solution because it can lead to inconsistent state.  For example, you might render a component with the previous state while another batched state change is in a new state.  There might be a nested component that then sees this mix of previous and next state which might be incompatible and lead to a bug.  Definitely came across that while using that hack and it can be annoying to debug.  That's part of what is nice about letting React Transitions handle the state changes consistently for branches of the component tree.

Think the previous value in a loadable would have any significant usability win over just using something like `usePrevious()`?

**Reply by drarmstr** [reply]
@salvoravida Nifty dev tools!

**Reply by salvoravida** [reply]
Yes, definitively, I think that a Loadable (that can be refreshed, think to fetch query) should have 4 states:
1 state = loading , contents:undefined,  -> refresh action -> remain state 1 - do nothing or reloading  (I guess current behavior is to recreate the promise)
2 state = hasValue, contents: Value. -> refresh action -> state 4
3 state = hasError, contents: Error -> refresh action -> state 1
4 state = isRefreshing, contents: Value -> refresh action -> do nothing state 4

this is very useful to not throw while refreshing; for example, if you want Silent refreshing, that is very common for many use cases.
(react-query use this approach)

another useful stuff would be adding a metadata object to the loadable for example, with a timestamp of the last result

2 state = hasValue, contents: Value, metadata: { timestamp: '....' }

**Reply by salvoravida** [reply]
@drarmstr I think that my solution of saving previous data in the ref is fine because I never use useRecoilQuery in nested components.

A typical use case is when you have the query used into 2 Widget already loaded, that shows the Value, and on the widget A,
the user clicks the button refresh: the behavior is a silent refresh on both widgets, with a loading bar on top using the updated loading status. without throwing stuff.

Anyway, I hope you will consider adding the 4 states "isRefreshing" so we have native support on the Loadable itself.
